### PR TITLE
test(numbers): pin #415 heterogeneous-comparison contract — variant ↔ std::integral witness tests + static_asserts

### LINKED ISSUE #415: feat(sets): SignedCardinality / Cardinality ↔ std::integral comparison operators
Sub-issue of epic #374. Filed 2026-04-26 from PR #412 / #402 prerequisite analysis.

**Blocks #402**  (re-scoped 2026-04-26 to cover both ℕ and ℤ) .

## Motivation

Discovered while attempting `using ℤ = SignedCardinality;` in PR #412: the halfspace machinery does

```cpp
constexpr Codomain operator()(const T& x) const {
  if constexpr (D == Direction::Upward) {
    const bool hit = (S == Strictness::Strict) ? (x > Pivot) : (x >= Pivot);
    ...
```

where `T` is the carrier (post-#402 retarget: `SignedCardinality`) and `Pivot` is the int NTTP from `bound<-21>`. After the retarget, `x > Pivot` becomes `SignedCardinality > int` — and `std::variant` provides no relational operators with primitives. Hard error inside the lambda body, propagated to every callsite that does `var<ℤ> > bound<-21>`.

Same applies to `Cardinality > unsigned int` once ℕ retargets too.

## Scope

Add explicit comparison operators in `sets:cardinality`:

```cpp
// SignedCardinality side
export template <std::integral S>
constexpr bool operator<(const SignedCardinality& lhs, S rhs) noexcept;
export template <std::integral S>
constexpr bool operator<=(const SignedCardinality& lhs, S rhs) noexcept;
export template <std::integral S>
constexpr bool operator>(const SignedCardinality& lhs, S rhs) noexcept;
export template <std::integral S>
constexpr bool operator>=(const SignedCardinality& lhs, S rhs) noexcept;
// (and reversed-LHS overloads for completeness)

// Cardinality side
export template <std::unsigned_integral U>
constexpr bool operator<(const Cardinality& lhs, U rhs) noexcept;
// ... etc.
```

Semantics: route via the existing variant-internal `compare(...)` helpers (`SignedCardinality vs SignedCardinality` already exists per `cardinality.cppm`'s arithmetic block); convert the int RHS to the appropriate variant alternative (the finite fragment) and delegate.

Edge cases to pin in tests:
- Comparison with `+ℵ_0` / `-ℵ_0`: any `int < +ℵ_0`, `int > -ℵ_0` regardless of sign.
- Comparison with `NaZ`: returns `false` for `<`, `<=`, `>`, `>=` (NaZ is incomparable, IEEE-NaN-style).
- Comparison with the finite fragment: standard signed/unsigned semantics.

## Acceptance criteria

- [ ] All four relational ops on `SignedCardinality × std::integral` compile and have specified semantics.
- [ ] Same for `Cardinality × std::unsigned_integral`.
- [ ] NaZ comparisons return false unanimously (not `Unknown` — these are bool returns; Ternary lift, if needed, is a separate concept).
- [ ] No regressions on existing 15/15 ctest + 136/136 Python tests.
- [ ] Witness tests in `signed_cardinality_test.cpp`: `SignedCardinality{7} > 5 == true`, `SignedCardinality{-7} > 5 == false`, `SignedCardinality{NaZ{}} > 5 == false`, etc.

## Out of scope

- `<=>` (spaceship) overloads — `bool` operators are sufficient for the halfspace use case; spaceship can be a follow-up.
- Comparisons between `Cardinality` and `SignedCardinality` directly — the embedding ℕ ↪ ℤ resolves cross-variant comparisons.

## Dependencies

- **Blocks** #402.
- **Independent of** SpeciesTraits specialisation (#413) and IsRingIntegral concept (#414).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by vincentk**
## Summary

Closes #415 (recon + contract pinning, not new operator surface).

The variant ↔ `std::integral` heterogeneous `<=>` overloads needed by #415 were already shipped in PR #431 (and earlier). This PR locks the contract with explicit witness tests so it stays frozen once #402 retargets `ℕ = Cardinality` and `ℤ = SignedCardinality`.

The #415 issue spec asked for explicit `operator<` / `<=` / `>` / `>=` overloads, but **C++20's spaceship-rewrite synthesises all four uniformly from the existing `<=>` overloads** — adding redundant single-purpose overloads would just be noise. The witness tests here document and freeze the working state instead.

## What lands

Two new test surfaces in [cardinality_test.cpp](src/test/cpp/modules/dedekind/numbers/cardinality_test.cpp):

### 1. Compile-time `static_assert` block (anonymous namespace at file scope)

~25 `static_assert`s covering both variant carriers vs signed / unsigned `std::integral`, in both forward and reversed directions:

- `Cardinality` × `unsigned` / `signed`: forward + reversed, including `ℵ_0` dominating.
- `SignedCardinality` × `signed` / `unsigned`: forward + reversed, including `±ℵ_0` dominating / dominated unconditionally.

These guarantee the comparisons remain **constexpr-evaluable** — the precondition for halfspace pivots like `bound<-21>` to keep working post-#402.

### 2. Runtime `TEST_CASE` for `SignedCardinality × std::integral`

The `Cardinality` side already had a #415-tagged `TEST_CASE` from PR #431; this adds the parallel `SignedCardinality` coverage:

- Positive finite vs signed and unsigned int.
- Negative finite vs signed and unsigned int (negative `SignedCardinality` < any `unsigned`).
- `±ℵ_0` dominates / is dominated unconditionally.
- `NaZ` propagates as unordered: all relational ops false, `==` false, `!=` true.
- `bool` integral path (no `make_unsigned_t<bool>` ill-formedness).
- Rhs-first direction synthesised by C++20 rewrite rules.

## Test results

64 assertions across 2 `TEST_CASE`s pass locally. No regressions on the rest of the test suite.

## Out of scope

- Single-purpose `operator<` / `<=` / `>` / `>=` overloads (redundant under spaceship).
- The `embed_K3_ℤ` retarget (#430) and `realize_signed_to_machine` extraction arrow (#429) — separate #402 prerequisites.
- The cascade sweep (#416) — depends on this contract being locked.

## References

- Issue #415 — heterogeneous comparison spec.
- PR #431 — original `<=>` overloads (`Cardinality / SignedCardinality × std::integral` and `× std::floating_point`).
- Issue #402 — umbrella for the variant-carrier retarget; this PR is one of the prerequisites.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/test/cpp/modules/dedekind/numbers/cardinality_test.cpp [hunk 1]
```diff
@@ -222,6 +222,120 @@ TEST_CASE(
   }
 }
 
+// Compile-time contract pinning for #415: variant carrier ↔ std::integral
+// relational ops MUST be constexpr-evaluable.  These static_asserts lock
+// the contract that downstream halfspace machinery relies on (e.g.\
```
[the remark this answers is not in the document]
Comment (vincentk) [reply src/test/cpp/modules/dedekind/numbers/cardinality_test.cpp]: Fixed in ccae161. Trailing backslash on the // comment was triggering Clang's -Wcomment 'multi-line // comment' warning, which would fail CI under -Werror. Rewrote the comment to drop the line-splicing backslash; no semantic change.

File: src/test/cpp/modules/dedekind/numbers/cardinality_test.cpp [hunk 2]
```diff
@@ -222,6 +222,120 @@ TEST_CASE(
...
+static_assert(sa_z_neg3 < 0);
+static_assert(sa_z_neg3 > -10);
+static_assert(sa_z_neg3 == -3);
+// ±ℵ_0 dominates any finite int unconditionally.
+static_assert(sa_z_pos_inf > 1'000'000);
+static_assert(sa_z_neg_inf < -1'000'000);
+// Reversed direction.
+static_assert(-3 < sa_z5);
+static_assert(0 > sa_z_neg3);
+static_assert(5 == sa_z5);
```
[the remark this answers is not in the document]
Comment (vincentk) [reply src/test/cpp/modules/dedekind/numbers/cardinality_test.cpp:281]: Fixed in ccae161. Extended the SignedCardinality static_assert block to mirror the runtime TEST_CASE: unsigned-int axis (sa_z5 > 4u, sa_z_neg3 < 0u as the negative-SignedCardinality-vs-unsigned witness, sa_z_neg3 < 5u), bool axis (sa_z5 > true/false, finite_signed_cardinality(1) == true), +ℵ_0 vs numeric_limits<unsigned>::max(), and rhs-first (4u < sa_z5, 5u == sa_z5, false < sa_z5). Compile-time and runtime surfaces now match exactly across both variant carriers.
